### DISCUSSION TITLE
Week View

### DISCUSSION THREAD

**Discussion by Lednerb** [wdelfuego/nova-calendar#18] [Ideas & Feature Requests]
As we already have the start and optionally also the end date and time of an event it would be very cool to have a view toggle button, to switch from the monthly view to a weekly one like it's integrated in thunderbird and many other tools:

![grafik](https://user-images.githubusercontent.com/2056876/174336840-2bd4b9b1-8416-403e-be9a-ffae4d3a7c36.png)

Therefore two options for the frontend might be a nice to have addition:
- Toggle between "show workdays" and "show all days" (incl. weekend)
- Set custom working hours for having them in the viewport (without scrolling)

What do you think about it?

NB: I like the clean and usable design and layout a lot, by implementing this Feature Request one should try to archive this to fit the current look and feel.

**Comment by wdelfuego** [comment]
A week view would definitely be nice to have. Two layout issues that will require smart UX decisions and possibly creative use of css grid features are overlapping time ranges for events and the display of multi-day events. The month calendar that we have now is 100% css, no JavaScript is used to layout the calendar events. It should be  possible to do the same thing for a week view but it won’t be trivial. 

I think a proper responsive interface for the month view so it can be used perfectly on mobile devices should have higher development priority and is also quite feasible, since it can use the same grid view we already have, but after that I can see a week view being a valuable addition to work on.

I’m curious, what is the specific use case that you would need a week view for? I understand that it makes sense for any calendar to include a week view but I like to question things we take for granted like that  because the answers might lead us in surprising directions.

**Reply by Lednerb** [reply]
Regarding the actual use-case:

I'm developing a custom all-in-one CRM, project management and file managing tool for my company.
We have different Nova Models like `Client`, `Task`, `Invoice` etc. and `Appointment`.

The `Appointment`s are either e.g. meetings with a specific `start` and `end` date time on the same day or are a multi-day event (exact hours do not matter for us in this case).
Models like `Task` or `Invoice` are just "Tasks" that have to be handled on that day but are not time-related, so it's sufficient for us to just display them in the day column.

We are trying to use the nova-calendar as a replacement for traditional calendar tools because we can easily integrate our event-related Nova models here without adding the same events on multiple places (paper, calendar tool, custom CRM solution, etc.).


![grafik](https://user-images.githubusercontent.com/2056876/174431381-df5977c5-034a-4d7c-a298-514860cfc827.png)

As you can see in the example screenshot, we use the `name` and `notes` functions for customizing the events.

With a weekly view it would be easier to find free time slots when you are in a phone call or meeting instead of calculating the free slots on a specific day by using just the brain ;-)

-----

In deed there are some more problems implementing the weekly view in a general manner like overlapping and concurrent events:

![grafik](https://user-images.githubusercontent.com/2056876/174431962-267ebe87-f0fb-4111-a0d9-f1de762a521b.png)


-----

It's a good approach to collect use-cases in order to find the best option for implementation and to decide if this feature should be implemented at all.

Thanks for your efforts!

**Reply by wdelfuego** [reply]
Thanks for the clear and detailed input, very valuable!

I can see this view being added at some point, but unless someone else implements it or some company is willing to sponsor its development I think it will be at least 6 months from now. 

Thanks a lot for the input, and please reserve more than 10 minutes for Frühstück ;).

**Reply by bb140856** [reply]
Hi Team,

I'm working on weekly view. In next steps: on daily as well as timeline views to allow user to choose timeslot to book the appointment. I'm happy to hear, that somebody is interested in it.

**Comment by wdelfuego** [comment]
@bb140856 Nice to hear that you're trying to implement this view! I'd like share my thoughts about it with you, for what it's worth.

I think for this to work elegantly, we need to split the `DataProvider\MonthCalendar` class into a superclass that holds generic dataprovider functionality and a `MonthCalendar` subclass that does everything that's specific to the current month calendar views, then add a second WeekCalendar subclass for this feature.

I think you can almost duplicate the way `Tool.vue` is currently implemented, just adapted to render the week view i/o the month view.

Both views should 100% share the CSS and event data structures from the back-end. If that means adding a field or two to all events for the week view, that's fine.

It's quite an undertaking for a big feature that has a lot of interaction with existing code and features, so it won't be trivial and I'll review critically before accepting pull requests. I'll be happy to offer guidance and feedback along the way to help things move in the right direction, so shoot me a message for any doubts you may have!

**Comment by bb140856** [comment]
@wdelfuego Thanks for support! Regarding to DataProvider, this is how I planned to do this. This is critical to ensure backward compatibility for MonthCalendar. We'll be in touch.

**Reply by wdelfuego** [reply]
Awesome 👍 .

Don't forget to check the [license model](https://github.com/wdelfuego/nova-calendar#license)!

**Reply by bb140856** [reply]
I've just created Pull request regarding to this feature. I've added 3 new views:
- Timeline
- Daily
- Weekly

Regarding to the DataProvider class - finally I've not divided into super- and subclasses due to archotectural reasons - it's injection approach. Next weeks im going to vacation so during this period you/wee can rethink is to continue one DataProvider, or divide it into few independent classes.

**Reply by wdelfuego** [reply]
Looks like a lot of effort, @bb140856 ! 👏 

Thanks for your pull request.
I will be reviewing it and get back to you with some comments soon; I haven't been able to see the features in action yet.

**Reply by jeremiahsherrill** [reply]
Any Update on this?

**Reply by wdelfuego** [reply]
@jeremiahsherrill , not for now, 2.0 is almost out and has some infrastructural preparations for the feature though. I'll be in touch with @bb140856 to discuss how he sees it.

**Comment by wdelfuego** [comment]
This feature is coming in v2.2, we're working on it on the dev branch now, latest info in [this pull request](https://github.com/wdelfuego/nova-calendar/pull/62)
